src/tools/simeng/main.cc

```diff
@@ -182,109 +150,42 @@ int main(int argc, char** argv) {
-      uint8_t port = reservation_station["Ports"][j].as<uint8_t>();
-      if (rsArrangement.size() < port + 1) {
-        rsArrangement.resize(port + 1);
-      }
-      rsArrangement[port] = {i, reservation_station["Size"].as<uint16_t>()};
-    }
+  if (mode == simeng::SimulationMode::InOrderPipelined) {
+    modeString = "In-Order Pipelined";
   }
 
-  int iterations = 0;
+  // Create memory interfaces
+  std::shared_ptr<simeng::MemoryInterface> dataMemory =
+      coreInstance->createL1DataMemory(L1Dtype);
+  std::shared_ptr<simeng::MemoryInterface> instructionMemory =
+      coreInstance->createL1InstructionMemory(simeng::MemInterfaceType::Flat);
 
-  // Dereferenced shared_ptr instance of process image held inside
-  // LinuxProcess.hh is passed to data memory.
-  std::shared_ptr<char> procImgForDataMem = process->getProcessImage();
-  char* processMemory = procImgForDataMem.get();
-  std::string modeString;
-  std::unique_ptr<simeng::Core> core;
-  std::unique_ptr<simeng::MemoryInterface> dataMemory;
-  switch (mode) {
-    case SimulationMode::OutOfOrder: {
-      modeString = "Out-of-Order";
-      dataMemory = std::make_unique<simeng::FixedLatencyMemoryInterface>(
-          processMemory, processMemorySize,
-          config["L1-Cache"]["Access-Latency"].as<uint16_t>());
-      core = std::make_unique<simeng::models::outoforder::Core>(
-          instructionMemory, *dataMemory, processMemorySize, entryPoint, *arch,
-          predictor, portAllocator, rsArrangement, config);
-      break;
-    }
-    case SimulationMode::InOrderPipelined: {
-      modeString = "In-Order Pipelined";
-      std::unique_ptr<simeng::FlatMemoryInterface> flatDataMemory =
-          std::make_unique<simeng::FlatMemoryInterface>(processMemory,
-                                                        processMemorySize);
-      core = std::make_unique<simeng::models::inorder::Core>(
-          instructionMemory, *flatDataMemory, processMemorySize, entryPoint,
-          *arch, predictor);
-      dataMemory = std::move(flatDataMemory);
-      break;
-    }
-    default: {
-      modeString = "Emulation";
-      dataMemory = std::make_unique<simeng::FlatMemoryInterface>(
-          processMemory, processMemorySize);
-      core = std::make_unique<simeng::models::emulation::Core>(
-          instructionMemory, *dataMemory, entryPoint, processMemorySize, *arch);
-      break;
-    }
-  };
-
-  simeng::SpecialFileDirGen SFdir = simeng::SpecialFileDirGen(config);
-  // Create the Special Files directory if indicated to do so in Config
-  if (config["CPU-Info"]["Generate-Special-Dir"].as<std::string>() == "T") {
-    // Remove any current special files dir
-    SFdir.RemoveExistingSFDir();
-    // Create new special files dir
-    SFdir.GenerateSFDir();
-  }
+  // Create core
+  std::shared_ptr<simeng::Core> core = coreInstance->createCore();
 
+  // Create Special Files directory if indicated to do so in Config
+  coreInstance->createSpecialFileDirectory();
```

Comment:
Could the coreInstance not call this itself?

Reply:
I can't remember which one of the interns had this, but they needed to not create the special file directory for some reason. Implemented it for scenarios equivalent to that.

Reply:
It may have been something to do with it being created in the default install directory which they don't have access to so maybe we could remove it as that's no the case anymore. Not sure if you can recall the issue better?

Reply:
This functionality can be set inside the configuration file already

Reply:
And yes, as the creation directory has been changed there should not be any access issues anymore

Reply:
I'll move out of its own function then.